File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation2/observables/geocode/ReverseGeocodeObservable.java
package pl.charmas.android.reactivelocation2.observables.geocode;

import android.content.Context;
import android.location.Address;
import android.location.Geocoder;

import java.io.IOException;
import java.util.List;
import java.util.Locale;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.ObservableEmitter;
import io.reactivex.rxjava3.core.ObservableOnSubscribe;
import io.reactivex.rxjava3.schedulers.Schedulers;
import pl.charmas.android.reactivelocation2.observables.ObservableEmitterWrapper;
import pl.charmas.android.reactivelocation2.observables.ObservableFactory;

public class ReverseGeocodeObservable implements ObservableOnSubscribe<List<Address>> {
    private final Context ctx;
    private final Locale locale;
    private final double latitude;
    private final double longitude;
    private final int maxResults;

    public static Observable<List<Address>> createObservable(Context ctx, ObservableFactory factory, Locale locale, double latitude, double longitude, int maxResults) {
        return factory.createObservable(new ReverseGeocodeObservable(ctx, locale, latitude, longitude, maxResults));
    }

    private ReverseGeocodeObservable(Context ctx, Locale locale, double latitude, double longitude, int maxResults) {
        this.ctx = ctx;
        this.latitude = latitude;
        this.longitude = longitude;
        this.maxResults = maxResults;
        this.locale = locale;
    }

    @Override
    public void subscribe(ObservableEmitter<List<Address>> emitter) throws Exception {
        Geocoder geocoder = new Geocoder(ctx, locale);
        try {
            List<Address> addresses = geocoder.getFromLocation(latitude, longitude, maxResults);
            if (!emitter.isDisposed()) {
                emitter.onNext(addresses);
                emitter.onComplete();
            }
        } catch (IOException e) {
            // If it's a service not available error try a different approach using google web api
            if (!emitter.isDisposed()) {
                Observable
                        .create(new FallbackReverseGeocodeObservable(locale, latitude, longitude, maxResults))
                        .subscribeOn(Schedulers.io())
                        .subscribe(new ObservableEmitterWrapper<>(emitter));
            }
        }
    }
}

File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation2/observables/geocode/ReverseGeocodeObservable.kt
package pl.charmas.android.reactivelocation2.observables.geocode

import android.content.Context
import android.location.Address
import android.location.Geocoder
import io.reactivex.Maybe
import io.reactivex.MaybeEmitter
import io.reactivex.MaybeOnSubscribe
import io.reactivex.schedulers.Schedulers
import pl.charmas.android.reactivelocation2.observables.MaybeEmitterWrapper
import pl.charmas.android.reactivelocation2.observables.MaybeFactory
import java.io.IOException
import java.util.Locale

class ReverseGeocodeObservable private constructor(
    private val ctx: Context,
    private val locale: Locale,
    private val apiKey: String,
    private val latitude: Double,
    private val longitude: Double,
    private val maxResults: Int
) : MaybeOnSubscribe<List<Address>> {

    override fun subscribe(emitter: MaybeEmitter<List<Address>>) {
        val geoCoder = Geocoder(ctx, locale)
        try {
            val addresses = geoCoder.getFromLocation(latitude, longitude, maxResults)
            if (!emitter.isDisposed) {
                emitter.onSuccess(addresses)
            }
        } catch (e: IOException) {
            // If it's a service not available error try a different approach using google web api
            if (!emitter.isDisposed) {
                Maybe.create(
                    FallbackReverseGeocodeObservable(
                        apiKey,
                        locale,
                        latitude,
                        longitude,
                        maxResults
                    )
                )
                    .subscribeOn(Schedulers.io())
                    .subscribe(MaybeEmitterWrapper(emitter))
            }
        }
    }

    companion object {
        fun create(
            ctx: Context,
            apiKey: String,
            factory: MaybeFactory,
            locale: Locale,
            latitude: Double,
            longitude: Double,
            maxResults: Int
        ): Maybe<List<Address>> {
            return factory.create(
                ReverseGeocodeObservable(ctx, locale, apiKey, latitude, longitude, maxResults)
            )
        }
    }
}